### PR TITLE
feat: add gate wallet adapter

### PR DESCRIPTION
**Pull request by weixuefeng**

### CODE REVIEW

**Comment by zorro11639** [comment]
Thanks for your contribution. The PR will be merged after testing.

File: packages/adapters/gatewallet/LICENSE [hunk 1]
```diff
@@ -0,0 +1,20 @@
+The MIT License (MIT)
+Copyright (c) 2022-Present, tronprotocol
```
Comment (zorro11639) [inline packages/adapters/gatewallet/LICENSE]: Copyright should update to `web3-geek`

File: packages/adapters/gatewallet/README.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        solidityNode: string;
+        eventServer: string;
+    };
+    ```
+
+### Caveats
+
+-   GateWallet App and Extension doesn't implement `multiSign()` and `switchChain()`.
+
+For more information about tronwallet adapters, please refer to [`@tronweb3/tronwallet-adapters`](https://github.com/tronprotocol/tronwallet-adapter/tree/main/packages/adapters/adapters)
```
Comment (zorro11639) [inline packages/adapters/gatewallet/README.md]: github repo url should update to https://github.com/web3-geek/tronwallet-adapter/tree/main/packages/adapters/adapters

File: packages/adapters/gatewallet/package.json [hunk 3]
```diff
@@ -0,0 +1,62 @@
+{
+    "name": "@tronweb3/tronwallet-adapter-gatewallet",
+    "version": "1.0.0",
+    "description": "Wallet adapter for Gate Wallet extension and Gate Wallet app.",
+    "keywords": [
+        "TRON",
+        "TronWeb",
+        "Gate Wallet"
+    ],
+    "author": "tronprotocol",
```
Comment (zorro11639) [inline packages/adapters/gatewallet/package.json]: `author/repo-url/homepage/bugs-url`  should update to web3-geek

File: packages/adapters/gatewallet/src/utils.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import { isInMobileBrowser } from '@tronweb3/tronwallet-abstract-adapter';
+
+export function supportGateWallet() {
+    return !!(window.gatewallet && window.gatewallet.tronLink);
```
Comment (zorro11639) [inline packages/adapters/gatewallet/src/utils.ts:4]: Is this function valid? I test the demo in GateWallet Ios app and browser with gate extension, but there is no `window.gatewallet.tronLink` .

File: packages/adapters/gatewallet/src/utils.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+}
+
+export const isGateApp = /GateApp/i.test(navigator.userAgent);
+
+export function openGateWallet() {
+    if (!isGateApp && isInMobileBrowser()) {
+        window.location.href =
```
Comment (zorro11639) [inline packages/adapters/gatewallet/src/utils.ts:11]: It seems that this deeplink format can not open the dapp in gatewallet ios mobile.

**Review by zorro11639 — COMMENTED** [review]
Hi, there are some issues to be fixed. Could you figure out the problems ?

File: packages/adapters/gatewallet/src/utils.ts (hunk 4 above)
Comment (zorro11639) [inline packages/adapters/gatewallet/src/utils.ts:4]: This function worked in mobile app but it seems strange. The first time I visit the website, I can not get `window.gatewallet.tronLink`. But after i click refresh button on the top right once or more, the `window.gatewallet.tronLink` will be valid. Is this a normal case ?

File: packages/adapters/gatewallet/src/utils.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+            'https://gateio.onelink.me/DmA6/web3?dapp_url=' + encodeURIComponent(window.location.href);
```
Comment (zorro11639) [inline packages/adapters/gatewallet/src/utils.ts:12]: I could open gatewallet app with this deeplink but it just open the home page not the dapp browser. Is this a normal case ?

File: packages/adapters/gatewallet/src/utils.ts (hunk 4 above)
Comment (weixuefeng) [reply packages/adapters/gatewallet/src/utils.ts:4]: Yes. it's my application problem. not adapter.  i'll check it.

File: packages/adapters/gatewallet/src/utils.ts (hunk 6 above)
Comment (weixuefeng) [reply packages/adapters/gatewallet/src/utils.ts:12]: had fixed

**Comment by zorro11639** [comment]
There are some issues to resolve. Could you have a look ?
1. **[IOS]**  Wallet imported using `privateKey` or `keystore` can not support DApp. Is this a deliberate design ?
2. **[Android]** Dapp can not detect `window.gatewallet` on Android gate.io app.

**Comment by weixuefeng** [comment]
> There are some issues to resolve. Could you have a look ?
> 
> 1. **[IOS]**  Wallet imported using `privateKey` or `keystore` can not support DApp. Is this a deliberate design ?
> 2. **[Android]** Dapp can not detect `window.gatewallet` on Android gate.io app.

1. if you import privateKey or keystore , you should  choose account to `Tron`, then it will support DApp.
2. maybe it is version problems? please use the latest version. https://www.gate.io/zh/mobileapp, download apk from website. and check update. new version had test.

**Comment by weixuefeng** [comment]
> > There are some issues to resolve. Could you have a look ?
> > 
> > 1. **[IOS]**  Wallet imported using `privateKey` or `keystore` can not support DApp. Is this a deliberate design ?
> > 2. **[Android]** Dapp can not detect `window.gatewallet` on Android gate.io app.
> 
> 1. if you import privateKey or keystore , you should  choose account to `Tron`, then it will support DApp.
> 2. maybe it is version problems? please use the latest version. https://www.gate.io/zh/mobileapp, download apk from website. and check update. new version had test.

image

**Comment by zorro11639** [comment]
I have download the latest Android from https://www.gate.io/zh/mobileapp. The version of Exchange is `v6.30.10` and the web3 is `v1.15.0`. But it does not work. 
1. Change the network to Tron with a `privateKey` wallet, it can not open Dapp browser.
2. Change the wallet to a Mnemonic wallet, it can open Dapp browser but no `window.gatewallet`.

I cannot find the reason. It seems strange.

**Comment by zorro11639** [comment]
I have download the latest Android from https://www.gate.io/zh/mobileapp. The version of Exchange is `v6.30.10` and the web3 is `v1.15.0`. But it does not work. 
1. Change the network to Tron with a `privateKey` wallet, it can not open Dapp browser.
2. Change the wallet to a Mnemonic wallet, it can open Dapp browser but no `window.gatewallet`.

I cannot find the reason. It seems strange.imageimage

**Comment by weixuefeng** [comment]
i'll fix it.

**Comment by weixuefeng** [comment]
> I have download the latest Android from https://www.gate.io/zh/mobileapp. The version of Exchange is `v6.30.10` and the web3 is `v1.15.0`. But it does not work.
> 
> 1. Change the network to Tron with a `privateKey` wallet, it can not open Dapp browser.
> 2. Change the wallet to a Mnemonic wallet, it can open Dapp browser but no `window.gatewallet`.
> 
> I cannot find the reason. It seems strange.imageimage

fixed。 Please update android version to 6.31.0. web3 version is 1.16.0. you can update android version in app. （check update）

**Comment by zorro11639** [comment]
Thanks for your work. I have checked the Android app. Now when i use the `privateKey` wallet, it support open the Dapp browser but there is still no `window.gatewallet`. @weixuefeng

**Comment by weixuefeng** [comment]
> Thanks for your work. I have checked the Android app. Now when i use the `privateKey` wallet, it support open the Dapp browser but there is still no `window.gatewallet`. @weixuefeng

thanks for your reply. 
How can i reproduction it? 
when i make demo for `window.gatewallet.tronLink`, it works. 
img_v3_02et_b8a3777b-e0b6-4139-9d0e-fd18db4442ix

**Comment by zorro11639** [comment]
We have test in Android 10 and Harmony OS 3.0. And the latest version is upgrade inside the App. But it does not work fine. What the OS and the version are you using ? Is your app downloaded from App store or gate.io website ?  Maybe you can provide a `apk` file which works fine on your device. @weixuefeng

**Comment by weixuefeng** [comment]
> We have test in Android 10 and Harmony OS 3.0. And the latest version is upgrade inside the App. But it does not work fine. What the OS and the version are you using ? Is your app downloaded from App store or gate.io website ? Maybe you can provide a `apk` file which works fine on your device. @weixuefeng

i download apk from website: https://www.gate.io/zh/mobileapp and upgrade inside the app. OS is  Android 12. i'll test & fix it.

**Comment by zorro11639** [comment]
It seems that the Android app works fine overall.
The latest test result:
- **[IOS]**
    - wallet imported by keystore can open dapp but it prompts "Not Support" when connect wallet in dapp
- **[Android]**
    - It works fine in some device such as Xiaomi Android12 and Huawei p50.
    - In some old device it still doesn't work fine Huawei P20-Android10 & Honor 20-android10
    - Can not import wallet by keystore

**Comment by weixuefeng** [comment]
> It seems that the Android app works fine overall. The latest test result:
> 
> * **[IOS]**
>   
>   * wallet imported by keystore can open dapp but it prompts "Not Support" when connect wallet in dapp
> * **[Android]**
>   
>   * It works fine in some device such as Xiaomi Android12 and Huawei p50.
>   * In some old device it still doesn't work fine Huawei P20-Android10 & Honor 20-android10
>   * Can not import wallet by keystore

When you import keystore. Only support evm network now. it's feature.
For old device, Current have not adapter it. can you merge it first?

**Comment by weixuefeng** [comment]
> > It seems that the Android app works fine overall. The latest test result:
> > 
> > * **[IOS]**
> >   
> >   * wallet imported by keystore can open dapp but it prompts "Not Support" when connect wallet in dapp
> > * **[Android]**
> >   
> >   * It works fine in some device such as Xiaomi Android12 and Huawei p50.
> >   * In some old device it still doesn't work fine Huawei P20-Android10 & Honor 20-android10
> >   * Can not import wallet by keystore
> 
> thanks for your reply. When you import keystore. Only support evm network now. it's feature. For old device, Current have not adapter it. can you merge it first? i'll test it on old devices and fix it.

**Comment by weixuefeng** [comment]
> > > It seems that the Android app works fine overall. The latest test result:
> > > 
> > > * **[IOS]**
> > >   
> > >   * wallet imported by keystore can open dapp but it prompts "Not Support" when connect wallet in dapp
> > > * **[Android]**
> > >   
> > >   * It works fine in some device such as Xiaomi Android12 and Huawei p50.
> > >   * In some old device it still doesn't work fine Huawei P20-Android10 & Honor 20-android10
> > >   * Can not import wallet by keystore
> > 
> > 
> > thanks for your reply. When you import keystore. Only support evm network now. it's feature. For old device, Current have not adapter it. can you merge it first? i'll test it on old devices and fix it. make pull request later.

**Comment by weixuefeng** [comment]
> > > > It seems that the Android app works fine overall. The latest test result:
> > > > 
> > > > * **[IOS]**
> > > >   
> > > >   * wallet imported by keystore can open dapp but it prompts "Not Support" when connect wallet in dapp
> > > > * **[Android]**
> > > >   
> > > >   * It works fine in some device such as Xiaomi Android12 and Huawei p50.
> > > >   * In some old device it still doesn't work fine Huawei P20-Android10 & Honor 20-android10
> > > >   * Can not import wallet by keystore
> > > 
> > > 
> > > thanks for your reply. When you import keystore. Only support evm network now. it's feature. For old device, Current have not adapter it. can you merge it first? i'll test it on old devices and fix it.

**Comment by zorro11639** [comment]
Thanks a lot. The pr is merged to release/v1.2.4 and we will update some dependencies. After testing, the gate adapter will be release in two days.

**Comment by zorro11639** [comment]
`@tronweb3/tronwallet-adapter-gatewallet` package is published. But when open gate.io app by deeplink, a dialog that prompts swiching wallet will show on some new Android devices. You could test some more devices to ensure the user experience.

**Comment by zorro11639** [comment]
Hi, the latest ios app(web3 v1.18.1) can not connect Tron wallet. Is this a normal case?

**Comment by weixuefeng** [comment]
> Hi, the latest ios app(web3 v1.18.1) can not connect Tron wallet. Is this a normal case?

which website? or how can i test it? @zorro11639

**Comment by zorro11639** [comment]
https://tronwallet-adapter-zeta.vercel.app/ you can visit this demo page and change to gatewallet tab. In my ios device, it can not connect the account. @weixuefeng

**Comment by weixuefeng** [comment]
> https://tronwallet-adapter-zeta.vercel.app/ you can visit this demo page and change to gatewallet tab. In my ios device, it can not connect the account. @weixuefeng

i can not visit the webite. can you provide the source code ? @zorro11639

**Comment by zorro11639** [comment]
Source code is in this repo, `demos/dev-demo` . Just run a js and console the `window.gatewallet` object and it doesn't contain `tronLink` object. @weixuefeng

**Comment by weixuefeng** [comment]
> Source code is in this repo, `demos/dev-demo` . Just run a js and console the `window.gatewallet` object and it doesn't contain `tronLink` object. @weixuefeng

ok, you can use latest version. new web3 version is 1.19.0. you can update it on appstore. @zorro11639

**Comment by zorro11639** [comment]
> > Source code is in this repo, `demos/dev-demo` . Just run a js and console the `window.gatewallet` object and it doesn't contain `tronLink` object. @weixuefeng
> 
> 
> 
> ok, you can use latest version. new web3 version is 1.19.0. you can update it on appstore. @zorro11639 

I cannot find the APP in the App Store . Is it available currently?

**Comment by weixuefeng** [comment]
> > > Source code is in this repo, `demos/dev-demo` . Just run a js and console the `window.gatewallet` object and it doesn't contain `tronLink` object. @weixuefeng
> > 
> > 
> > ok, you can use latest version. new web3 version is 1.19.0. you can update it on appstore. @zorro11639
> 
> I cannot find the APP in the App Store . Is it available currently?

maybe you should use appleId in HK

**Pull request merged**